libmscore/score.cpp

```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
-                  // special case for initial "C" key signature - these have no explicit element
-                  if (!seg->element(staffIdx * VOICES)) {
-                        // no need to create new initial "C" key sig
-                        // if staff already ends in that key
-                        if (joinedStaff->key(tickOfAppend - Fraction::fromTicks(1)) == Key::C)
-                              continue;
-                        Key key = Key::C;
-                        KeySig* ks = new KeySig(this);
-                        ks->setTrack(staffIdx * VOICES);
-                        ks->setKey(key);
-                        ks->setParent(seg);
-                        addElement(ks);
-                        }
-                  // other key signatures (initial other than "C", non-initial)
-                  for (auto k : *(st->keyList())) {
-                        Fraction tick = Fraction::fromTicks(k.first);
-                        KeySigEvent key = k.second;
-                        joinedStaff->setKey(tick + tickOfAppend, key);
-                        }
+      // at first added measure, check if we need to add Clef/Key/TimeSig
+      //  this is needed if it was changed and needs to be changed back
+      int n = nstaves();
+      Fraction otick = fmb->tick(), ctick = tickOfAppend;
+      qDebug() << "Checking if something needs to be added" << otick << ctick;
+      for (int staffIdx = 0; staffIdx < n; ++staffIdx) { // iterate over all staves
+            int trackIdx = staff2track(staffIdx); // idx of irst track on the staff
+            Staff* staff = this->staff(staffIdx);
+            Staff* ostaff = score->staff(staffIdx);
+            
+            // check if key signature needs to be changed
+            if (!(ostaff->key(otick) == staff->key(ctick))) {
+
+                  Segment* ns = firstAppendedMeasure->undoGetSegment(SegmentType::KeySig, ctick);
+                  KeySigEvent nkse = KeySigEvent(ostaff->keySigEvent(otick));
+                  KeySig* nks = new KeySig(this);
+                  nks->setScore(this);
+                  nks->setTrack(trackIdx);
+
+                  nks->setKeySigEvent(nkse);
+                  staff->setKey(ctick, nkse);
+                  ns->add(nks);
+                  //undoAddElement(nks);
+                  }
+            // check if a key signature is present but is spurious (i.e. no actual change)
+            else if (staff->currentKeyTick(ctick) == ctick &&
+                     staff->key(ctick - Fraction::fromTicks(1)) == ostaff->key(otick)) {
+                        Segment* ns = firstAppendedMeasure->first(SegmentType::KeySig);
+                        if (ns) ns->remove(ns->element(trackIdx));
+                  }
+            
+            // check if time signature needs to be changed
+            if (!ostaff->timeSig(otick) || !staff->timeSig(ctick)) {} // yes, this can happen.
+            else if (!((*ostaff->timeSig(otick)) == (*staff->timeSig(ctick)))) {
```

Comment:
Rather than this slightly confusing code, I think you should just make two variables and only one if statement. That way it’ll be immediately obvious as to what you are trying to do. Also, it will simplify some of the code later on.

Reply:
@JoshuaBonn1 I do not quite understand what you mean by "two variables" here. Could you elaborate a bit more?

Reply:
I realize that I made a mistake and I see what you are doing. Now it's on to personal preference of reducing the code a bit. However, unless it breaks something, I think the extra parenthesis around the `*ostaff->timeSig(otick)` and `*staff->timeSig(ctick)` are excessive.

What follows below is simply preference. You may ignore it.

`ostaff->timeSig(otick)` and `staff->timeSig(ctick)` can be made into two variables.
`TimeSig* ts = staff->timeSig(ctick);`
`TimeSig* ots = ostaff->timeSig(otick);`
`if (!ts || !ots) {}`
Then replace with `ts` and `ots` where possible later on in the code.

Reply:
Perhaps this is what Joshua was suggesting:
```
TimeSig* ts1 = ostaff->timeSig(otick);
TimeSig* ts2 = staff->timeSig(ctick);
if (ts1 && ts2 && (ts1->sig() != ts2->sig()) { 
```
I presume that it is not necessary to test for full equivalence between the two TimeSig objects, and that comparing their sig() values is enough.